From 3564954ec43d58471027322f20627f4a24c4b777 Mon Sep 17 00:00:00 2001
From: Raphael Mounier <mounierr07@gmail.com>
Date: Thu, 24 Feb 2022 14:53:03 +0100
Subject: [PATCH] Enable key authentification by property

Enable key authentification by setting, property ro.crypto.keyauth.enable to 1
---
 KeyStorage.cpp | 18 ++++++++++++++++++
 KeyStorage.h   |  7 ++++---
 2 files changed, 22 insertions(+), 3 deletions(-)

diff --git a/KeyStorage.cpp b/KeyStorage.cpp
index b64ddea..d11c0a8 100644
--- a/KeyStorage.cpp
+++ b/KeyStorage.cpp
@@ -81,6 +81,24 @@ static const char* kFn_secdiscardable = "secdiscardable";
 static const char* kFn_stretching = "stretching";
 static const char* kFn_version = "version";
 
+
+
+// Huawei patch by Iceows - Enable software authentification key
+KeyAuthentication::KeyAuthentication(const std::string& t, const std::string& s) : token{t}, secret{s}  {
+
+      if (s.empty()) {
+        char kacrypt_prop[PROPERTY_VALUE_MAX];
+
+        property_get("ro.crypto.keyauth.enable", kacrypt_prop, "");
+        if (strlen(kacrypt_prop) > 0 && std::stoi(kacrypt_prop)) {
+          // Set software hard code key
+          LOG(INFO) << "Enable key authentification - ro.crypto.keyauth.enable=1";
+          secret = "N0BZGwYKev0auvSYIocce4MB876RES8C";
+        } 
+      }
+}
+
+
 static bool checkSize(const std::string& kind, size_t actual, size_t expected) {
     if (actual != expected) {
         LOG(ERROR) << "Wrong number of bytes in " << kind << ", expected " << expected << " got "
diff --git a/KeyStorage.h b/KeyStorage.h
index 5228f08..2bdcc60 100644
--- a/KeyStorage.h
+++ b/KeyStorage.h
@@ -30,13 +30,14 @@ namespace vold {
 // binary needed to unlock.
 // If only "secret" is nonempty, it is used to decrypt in a non-Keymaster process.
 class KeyAuthentication {
+
   public:
-    KeyAuthentication(const std::string& t, const std::string& s) : token{t}, secret{s} {};
+    KeyAuthentication(const std::string& t, const std::string& s);
 
     bool usesKeymaster() const { return !token.empty() || secret.empty(); };
 
-    const std::string token;
-    const std::string secret;
+    std::string token;
+    std::string secret;
 };
 
 extern const KeyAuthentication kEmptyAuthentication;
-- 
2.30.2

